android/app/src/main/java/app/organicmaps/util/SpeedFormatted.java

```diff
@@ -0,0 +1,55 @@
+package app.organicmaps.util;
+
+import android.content.Context;
+
+import androidx.annotation.Keep;
+import androidx.annotation.NonNull;
+import androidx.annotation.StringRes;
+
+import app.organicmaps.R;
+
+// Used by JNI.
+@Keep
+@SuppressWarnings("unused")
+public class SpeedFormatted
+{
+  /**
+   * IMPORTANT : Order of enum values MUST BE the same as native measurement_utils::Units enum.
+   */
+  public enum Units
+  {
+    KilometersPerHour(R.string.kilometers_per_hour),
+    MilesPerHour(R.string.miles_per_hour);
+
+    @StringRes
+    public final int mStringRes;
+
+    Units(@StringRes int stringRes)
+    {
+      mStringRes = stringRes;
+    }
+  }
+
+  public final double mSpeed;
+  @NonNull
+  public final String mSpeedStr;
+  public final SpeedFormatted.Units mUnits;
+
+  public SpeedFormatted(double mSpeed, @NonNull String mSpeedStr, byte unitsIndex)
```

Comment:
nit: Are there simpler ideas on the implementation? E.g. return an array of 3 strings instead of introducing a class? How will it work on iOS? On Qt/desktop platforms?

Reply:
In iOS code navigation UI get speed from iOS API and speed limit from DTO `MWMNavigationDashboardEntity` (see [NavigationControlView.swift:159](https://github.com/organicmaps/organicmaps/blob/master/iphone/Maps/Classes/CustomViews/NavigationDashboard/Views/NavigationControlView.swift#L159) ). iOS converts m/s to km/h or mi/h in UI code calling `measurement_utils::MpsToUnits(...)`. Also it compares speed and limit values in m/s and doesn't do rounding.

To avoid JNI conversion from m/s to metric or imperial units I pass current speed and speed limit using `SpeedFormatted` class. So we have only one JNI call to `Framework.nativeGetRouteFollowingInfo()`.

You propose to pass `String speedStr`, `String speedLimitStr`, `byte speedUnits`, `bool isSpeedLimitExceeded` instead?

Reply:
On iOS information comes from two methods:
```
- (void)updateFollowingInfo:(routing::FollowingInfo const &)info routePoints:(NSArray<MWMRoutePoint *> *)points type:(MWMRouterType)type

and

- (void)updateTransitInfo:(TransitRouteInfo const &)info
```

It would be great to have one source of truth on all platforms, without any logic. E.g. just strings to display and flags like "speed limit was hit", or better 3-state like "no speeding/limit hit/fine limit hit".

To avoid slow JNI calls, something like this may be used:
`return currentSpeedStr.append(';').append(speedLimit).append(';').append(speedingFlag).apend(';').append(unitsStr);`

Then it can be decoded in Java right before displaying.

Passing all values separately is also ok, just slower a bit.

Reply:
As far as I can tell, iOS doesn't have problem with units and formatting. Swift calls C++ function from `measurement_utils` namespace to convert m/s to units-per-hour and format speed to string. No problems with Swift<->C++ calls, as far as I can see.

New approach when `MWMRoutingManager` for iOS and `Framework.nativeGetRouteFollowingInfo()` for Android do speed convertions and have some new protocol with `';'` separators looks like overengineering to me.

I propose to drop `SpeedFormatted` code and let UI do units conversions and string formatting with plain Java. I think functions within `measurement_utils` namespace would be easy to port to Java. This would minimize objects allocation between C++ and Java (get back `double speedLimitMps` as in `master`). What does iOS devs think of this?